### DISCUSSION TITLE
File "/home/runner/work/OK/OK/xray/xrayping.py", line 13, in <module>

### DISCUSSION THREAD

**Discussion by vpnineh** [community/community#138936] [Actions] [answered]
### Select Topic Area

Question

### Body

hello
why this error get back two username for the path?
thanks

**Comment by barishaxxer** [comment]
the error message you have provided is not complete.it looks like a standard directory structure possibly in a CI/CD environment (given the runner directory).
the relevant code around line 13 of xrayping.py and full error message would be good.
if you r using os.path.join() or a similar method to construct paths  ensure that you are not adding unnecessary components (like extra /home/username/) more than once. Since you're using a runner, check if the configuration file or workflow YAML might contain a mistake where paths are being incorrectly defined or duplicated. if this solves your problem please mark this comment as answer or if you have further questions let me know

**Reply by vpnineh** [reply]
Full Error : 
Run python xray/xrayping.py
Traceback (most recent call last):
  File "/home/runner/work/WOW/WOW/xray/xrayping.py", line 13, in <module>
    from modules.xray_url_decoder.XrayUrlDecoder import XrayUrlDecoder
  File "/home/runner/work/WOW/WOW/xray/modules/xray_url_decoder/XrayUrlDecoder.py", line 6, in <module>
    from .IsValid import isValid_tls, isValid_reality, isValid_userVless, isValid_vnextVless, isValid_link
  File "/home/runner/work/WOW/WOW/xray/modules/xray_url_decoder/IsValid.py", line 5, in <module>
    from .vless import UserVless, VnextVless
  File "/home/runner/work/WOW/WOW/xray/modules/xray_url_decoder/vless.py", line 5, in <module>
    from xray_url_decoder.XraySetting import StreamSettings, Mux
ModuleNotFoundError: No module named 'xray_url_decoder'
Error: Process completed with exit code 1.

**Reply by vpnineh** [reply]
**Worker :**

name: Ping Configs

on:
  push:
    paths:
      - '.github/workflows/xrayping.yml'
      - 'xray/xrayping.py'
      - 'xray/configs/all_configs.txt'
      - 'xray/modules/gitRepo.py'
  workflow_dispatch:

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2

    - name: Set up Python
      uses: actions/setup-python@v2
      with:
        python-version: '3.x'

    - name: Install dependencies
      run: |
        python -m pip install --upgrade pip
        pip install -r xray/requirements.txt

    - name: Run sorter script
      run: python xray/xrayping.py
      
    - name: Check for Output Changes
      id: xrays
      run: |
        git diff --quiet && echo 'No changes' || echo '::set-output name=changed::true'   
      
    - name: Commit and push sorter changes
      if: steps.xrays.outputs.changed == 'true'   
      run: |
        git config --local user.email "action@github.com"
        git config --local user.name "GitHub Action"
        git add xray/configs/
        git commit -m "Update files" -a
        git push


**Main Code :**


import sys
sys.path.append('./modules/xray_url_decoder/')
sys.path.append('./modules/clash_meta_url_decoder/')
sys.path.append('./')
sys.path.append('./modules')
sys.path.append('xray/modules/gitRepo.py')

import json
import uuid
from ruamel.yaml import YAML
from modules.gitRepo import commitPushRActiveProxiesFile, getLatestActiveConfigs

from modules.xray_url_decoder.XrayUrlDecoder import XrayUrlDecoder
from modules.XrayPing import XrayPing
from modules.clash_meta_url_decoder.ClashMetaUrlDecoder import ClashMetaDecoder


def is_good_for_game(config: XrayUrlDecoder):
    return (config.type in ['tcp', 'grpc']) and (config.security in [None, "tls"])


# for more info, track this issue https://github.com/MetaCubeX/Clash.Meta/issues/801
def is_buggy_in_clash_meta(config: ClashMetaDecoder):
    return config.security == "reality" and config.type == "grpc"




with open("xray/configs/raw-url/all.txt", 'r') as rowProxiesFile:
    configs = []
    clash_meta_configs = []
    for_game_proxies = []
    for url in rowProxiesFile:
        if len(url) > 10:
            try:
                cusTag = uuid.uuid4().hex

                # ############# xray ############
                c = XrayUrlDecoder(url, cusTag)
                c_json = c.generate_json_str()
                if c.isSupported and c.isValid:
                    configs.append(c_json)

                # ############# clash Meta ##########
                ccm = ClashMetaDecoder(url, cusTag)
                ccm_json = ccm.generate_obj_str()
                if c.isSupported and c.isValid and (not is_buggy_in_clash_meta(ccm)):
                    clash_meta_configs.append(json.loads(ccm_json))

                if is_good_for_game(c):
                    for_game_proxies.append(url)
            except:
                print("There is error with this proxy => " + url)

    # getLatestGoodForGame()
    # with open("xray/configs/row-url/for_game.txt", 'w') as forGameProxiesFile:
    #     for forGame in for_game_proxies:
    #         forGameProxiesFile.write(forGame)
    # commitPushForGameProxiesFile()

    delays = XrayPing(configs, 200)
    getLatestActiveConfigs()

    yaml = YAML()
    with open("xray/configs/clash-meta/all.yaml", 'w') as allClashProxiesFile:
        yaml.dump({"proxies": clash_meta_configs}, allClashProxiesFile)

    with open("xray/configs/clash-meta/actives_under_1000ms.yaml", 'w') as active1000ClashProxiesFile:
        values_to_filter = {d['proxy']['tag'].split("_@_")[0] for d in delays.realDelay_under_1000}
        filtered_array = [item for item in clash_meta_configs if item['name'].split("_@_")[0] in values_to_filter]
        yaml.dump({"proxies": filtered_array}, active1000ClashProxiesFile)

    with open("xray/configs/clash-meta/actives_under_1500ms.yaml", 'w') as active1500ClashProxiesFile:
        values_to_filter = {d['proxy']['tag'].split("_@_")[0] for d in delays.realDelay_under_1500}
        filtered_array = [item for item in clash_meta_configs if item['name'].split("_@_")[0] in values_to_filter]
        yaml.dump({"proxies": filtered_array}, active1500ClashProxiesFile)

    with open("xray/configs/xray-json/actives_all.txt", 'w') as activeProxiesFile:
        for active in delays.actives:
            activeProxiesFile.write(json.dumps(active['proxy']) + "\n")

    with open("xray/configs/xray-json/actives_under_1000ms.txt", 'w') as active1000ProxiesFile:
        for active in delays.realDelay_under_1000:
            active1000ProxiesFile.write(json.dumps(active['proxy']) + "\n")

    with open("xray/configs/xray-json/actives_under_1500ms.txt", 'w') as active1500ProxiesFile:
        for active in delays.realDelay_under_1500:
            active1500ProxiesFile.write(json.dumps(active['proxy']) + "\n")

    with open("xray/configs/xray-json/actives_no_403_under_1000ms.txt", 'w') as active1000no403ProxiesFile:
        for active in delays.no403_realDelay_under_1000:
            active1000no403ProxiesFile.write(json.dumps(active['proxy']) + "\n")

    with open("xray/configs/xray-json/actives_for_ir_server_no403_u1s.txt",
              'w') as active1000no403ForServerProxiesFile:
        for active in delays.no403_realDelay_under_1000:
            if active['proxy']["streamSettings"]["network"] not in ["ws", "grpc"]:
                active1000no403ForServerProxiesFile.write(json.dumps(active['proxy']) + "\n")




commitPushRActiveProxiesFile()

**Reply by barishaxxer** [reply, marked as answer]
This is likely due to an issue with the Python path or how the modules are being imported. first fix your imports as
 ```python
 import sys
import os

sys.path.append(os.path.dirname(os.path.abspath(__file__)))

from modules.xray_url_decoder.XrayUrlDecoder import XrayUrlDecoder
from modules.XrayPing import XrayPing
from modules.clash_meta_url_decoder.ClashMetaUrlDecoder import ClashMetaDecoder
from modules.gitRepo import commitPushRActiveProxiesFile, getLatestActiveConfigs
```
your directory structure looks like this i suppose
WOW/
├── xray/
│   ├── xrayping.py
│   ├── modules/
│   │   ├── __init__.py
│   │   ├── gitRepo.py
│   │   ├── XrayPing.py
│   │   ├── xray_url_decoder/
│   │   │   ├── __init__.py
│   │   │   ├── XrayUrlDecoder.py
│   │   │   ├── IsValid.py
│   │   │   ├── vless.py
│   │   │   └── XraySetting.py
│   │   └── clash_meta_url_decoder/
│   │       ├── __init__.py
│   │       └── ClashMetaUrlDecoder.py
│   └── configs/
│       └── all_configs.txt
└── .github/
    └── workflows/
        └── xrayping.yml
      ensure there is a   __init__.py file in each module directory, including the xray_url_decoder and clash_meta_url_decoder directories. These files can be empty, but they're necessary for Python to treat the directories as packages
      If you're using any third-party packages, make sure they're listed in your requirements.txt file and that the file is in the correct location (probably in the xray/ directory).
      fix xrayping yaml as - name: Run sorter script
  run: |
    cd xray
    python xrayping.py.
    this should solve your problem if dont. add something like this code to debug 


print("Python path:", sys.path)
print("Current working directory:", os.getcwd())


if problem persists send the requirements.txt files content.if solved please mark this comment as the answer ❤️
